File: ABC/ABC301-400/ABC364/F/event_sort/F.py
from sortedcontainers import SortedList

N, Q = map(int, input().split())
events = []
ans = 0
for _ in range(Q):
	L, R, C = map(int, input().split())
	events.append((L - 1, C, 0)) # 区間に入るイベント
	events.append((R - 1, C, 1)) # 区間から抜けるイベント
	ans += C
events.sort()
st = SortedList() # 今見ているインデックスでありえる C の集合
ind = 0
for i in range(N - 1):
	while ind < len(events) and events[ind][0] == i:
		if events[ind][2] == 0:
			st.add(events[ind][1])
		else:
			st.discard(events[ind][1])
		ind += 1
	if len(st) == 0:
		print(-1)
		exit()
	ans += st[0]
print(ans)

File: ABC/ABC301-400/ABC364/F/event_sort/F.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
	int N, Q;
	cin >> N >> Q;
	vector<tuple<int, ll, bool>> events;
	ll ans = 0;
	for (int i = 0; i < Q; i++) {
		ll C;
		int L, R;
		cin >> L >> R >> C;
		L--;
		R--;
		events.push_back({L, C, false});
		events.push_back({R, C, true});
		ans += C;
	}
	sort(events.begin(), events.end());
	int ind = 0;
	multiset<int> ms;
	for (int i = 0; i < N - 1; i++) {
		while (ind < events.size() && get<0>(events[ind]) == i) {
			if (!get<2>(events[ind])) {
				ms.insert(get<1>(events[ind]));
			} else {
				ms.erase(ms.find(get<1>(events[ind])));
			}
			ind++;
		}
		if (ms.empty()) {
			cout << -1 << '\n';
			return 0;
		}
		ans += *ms.begin();
	}
	cout << ans << '\n';
}